blockstore/apps/bundles/models.py

```diff
@@ -0,0 +1,203 @@
+    title = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_index=True)
+
+    def __str__(self):
+        return "{} - {}".format(self.uuid, self.title)
+
+
+class Bundle(models.Model):
+    """
+    The UUID of a Bundle should be treated as immutable, but any other part of
+    the bundle can be modified without breaking anything.
+    """
+    id = models.BigAutoField(primary_key=True)
+    uuid = models.UUIDField(default=uuid.uuid4, unique=True, editable=False)
+    title = models.CharField(max_length=MAX_CHAR_FIELD_LENGTH, db_index=True)
+
+    # Collections are not just an organizational tool, but also determine
+    # ownership, licensing, permissions, etc. So we expect to only belong to one
+    # Collection at any given time.
+    #
+    # Comment for now just to make testing easier.
+    collection = models.ForeignKey(
+        Collection, related_name="bundles", related_query_name="bundle"
+    )
+
+    slug = models.SlugField(allow_unicode=True)  # For pretty URLs
+    description = models.TextField(max_length=10_000)
+
+    def __str__(self):
+        return "Bundle {} - {}".format(self.uuid, self.slug)
+
+
+@attr.s(frozen=True)
+class BundleVersionKey():
+    bundle_uuid = attr.ib(type=uuid.UUID)
+    version_num = attr.ib(type=int)
+
+    @classmethod
+    def parse(cls, key_str):
+        """
+        Format for this is {UUID}@{version_num}.
+        """
+        try:
+            bundle_uuid_str, version_str = key_str.split('@')
+            bundle_uuid = uuid.UUID(hex=bundle_uuid_str)
+            version_num = int(version_str)
+        except Exception:
+            raise ValueError("Could not parse BundleVersionKey from {}".format(part_id_str))
+
+        return cls(bundle_uuid, version_num)
+
+    def __str__(self):
+        return "{}@{}".format(self.bundle_uuid, self.version_num)
+
+
+class BundleVersion(models.Model):
+    """
+
+    BundleVersions are almost like numbered tags.
+    """
+    id = models.BigAutoField(primary_key=True)
+    bundle = models.ForeignKey(
+        Bundle, related_name="versions", related_query_name="version"
+    )
+    version_num = models.PositiveIntegerField()
+
+    snapshot_digest = models.BinaryField(max_length=20, db_index=True)
+    # compatibility_hash = Let content determine a hash so that two versions can
+    #                      be compared for compatibility (ORA2 does this). A
+    #                      proxy for breaking change or major/minor versions?
+
+    change_description = models.TextField(max_length=1_000, blank=True)
+
+    @staticmethod
+    @receiver(snapshot_created)
+    def listen_for_snapshot_creation(**kwargs):
+        bundle_uuid = kwargs['bundle_uuid']
+        snapshot_digest = kwargs['snapshot_digest']
+
+        bundle = Bundle.objects.get(uuid=bundle_uuid)
+        versions = list(bundle.versions.order_by('-version_num')[:1])
+        next_version_num = versions[0].version_num + 1 if versions else 1
+
+        bundle.versions.create(
```

Comment:
In final point of the [Blockstore Design: General Themes/Concepts](https://openedx.atlassian.net/wiki/spaces/AC/pages/737149430/Blockstore+Design#Blockstore-Design-genera), it says: 

> versions are meaningful, and not every edit of every file spawns a new version. A version is like a "commit" in that sense.

It seems like creating versions automatically when new snapshots (files) get added to a bundle violates this statement?

Reply:
This is my fault for not making any clear sketch of how drafts should work.

I was thinking that snapshots represented a state of the file tree with all file data, like a commit in git. Snapshots and versions would mostly be 1:1 with each other, though there might be rollback situations where multiple versions point to the same snapshot. But snapshots were still meaningful in the sense that they represent a desired state of the Bundle's files and not an intermediate editing state.

For the intermediate editing state, I thought we'd have a directory of resources that was basically a copy-on-write (from the most recent BundleVersion) scratch space. And that when you hit "publish", Blockstore would take the contents of that scratch space and create a new snapshot from it. (Insert hand waving)

Reply:
> But snapshots were still meaningful in the sense that they represent a desired state of the Bundle's files and not an intermediate editing state.

Ok, I think I understand.. 

A "snapshot" is a deliberate action by the user, saying "I want to save this state", and so these should have annotations attached, and we'll provide a way for users to redo/undo between them.

But a "version" is every single edit that gets made to the bundle, and so it's automatically created, as shown above.

Is that right?

Reply:
The way I understand it is that a snapshot refers to the state of bundle content while the `Version` model is for keeping track of the order of snapshots of a bundle. What you are referring to by "(tracking) deliberate action by the user" is something that the `Version` model should keep track of. I think the later needs some more thinking. I have some thoughts/notes that I'll be posting later on the data model PR.

Reply:
@bradenmacdonald One complexity is that unlike in Studio, a bundle can be much larger than a Unit. What happens if two people are editing different parts of a bundle at the same time and one hits publish? 

Some possible options:

1. The API has restrictions on concurrent editing of the same bundle.
2. The API provides information about concurrent edits taking place so that the authoring interface can warn authors.
3. The API has support for multiple draft branches.
4. The API lets author deal with this themselves.

Are there any requirements from product regarding this?

Reply:
>My preference and the way I had been thinking about it is: that we have numbered versions like 1,2, 3, 4 and a draft version. The numbered versions are always immutable and only created explicitly by a "publish" action that the user takes. The draft version is a mutable version and is the working copy where changes can be made. Hitting publish creates a new numbered version that matches the draft version, and hitting "discard changes" reverts the mutable draft version to the latest numbered version if any. Hitting "revert to version N" produces a new numbered version identical to the specified older version.

I agree with how drafts work conceptually as a working copy, and the publish/discard/revert workflow you outline here. From the user's perspective, the draft can be something that is presented as a version-in-flight. But I think it should be modeled as something entirely separate from BundleVersions, since it differs so much in mutability, data representation, update notifications, associated metadata, discovery, etc.

> @bradenmacdonald One complexity is that unlike in Studio, a bundle can be much larger than a Unit. What happens if two people are editing different parts of a bundle at the same time and one hits publish?
>
>Some possible options:
>
>1. The API has restrictions on concurrent editing of the same bundle.
>2. The API provides information about concurrent edits taking place (including the diff of the draft from the last published version) so that the authoring interface can warn authors.
>3. The API has support for multiple draft branches.
>4. The API lets author deal with this themselves.

I think we leave the door open for (3) in the long term by not making Bundles aware of their drafts -- i.e. have the Drafts push into BundleVersions but don't have Bundles assume they have one and only one Draft. But (2) gets us 90% of the way towards what we need for a lot less complexity. A simple log showing who else is doing editing will let the authors coordinate amongst themselves, and course teams are small enough where I think that should be sufficient most of the time. Teams with really high concurrency needs can still use the github-push workflow that they probably already use.

Reply:
^ That all sounds good to me :)  I do think that (2) gives us the most value in the short term.

Reply:
^^ sounds good to me too, and letting humans negotiate to resolve concurrent edits is much better than trying to sort them out automatically.

But despite the fact that we've agreed that:
> The numbered versions are always immutable and only created explicitly by a "publish" action that the user takes. 

A new [`BundleVersion` is still auto-created when a `BundleSnapshot` is created](https://github.com/open-craft/blockstore/blob/master/blockstore/apps/bundles/models.py#L132-L150).  I would say that this should be resolved as part of a subsequent PR, but this REST API is reflecting this same confusion.

Instead, I would expect that:
* the REST API (and/or the models underneath) would create (only) a `BundleSnapshot`  when changes are made to a `Bundle` or its constituent `BundleFiles`.
  That is, neither the REST API nor the models should be auto-creating `BundleVersion` objects.
* the REST API would provide an explicit "Create `BundleVersion`" endpoint, that can be invoked when a human decides to label a particular snapshot with a version.
* the REST API would also provide a "Modify `BundleVersion`" endpoint, which allows users to edit the parts of the `BundleVersion` which are mutable -- currently just the `change_description`, but in the future, maybe metadata.
* the REST API would provide a way to list/view the `BundleSnapshot` list, so that `BundleVersion`s can be created from the desired snapshot.
* We'll figure out how to deal with Drafts later, since they're special.
  They are mutable, or perhaps, applying updates to a Draft will destroy previous the Snapshot.

@symbolist  @ormsbee  Am I missing something crucial here?

Reply:
I'm still not clear on why [saving a `BundleSnapshot` automatically triggers a new `BundleVersion` to be created](https://github.com/open-craft/blockstore/blob/master/blockstore/apps/bundles/models.py#L132-L150).  But it seems like I'm the only one who's confused about this, so it feels silly to keep blocking the review on this point.

So am letting this go for now.. though I expect that it'll come up later once we stick a GUI on this 😄